File: frontend/src/types/video-workflow.ts
/**
 * TypeScript types for Video Workflow feature.
 * Matches Pydantic schemas from backend.
 */

// ============================================================================
// Enums
// ============================================================================

export type WorkflowGoal =
  | "product_ad"
  | "ugc"
  | "explainer"
  | "cinematic_brand"
  | "tutorial"
  | "other"

export type WorkflowPlatform =
  | "tiktok"
  | "instagram"
  | "youtube"
  | "web"
  | "other"

export type AspectRatio = "9:16" | "1:1" | "16:9"

export type WorkflowStatus = "draft" | "in_progress" | "completed"

export type ShotType = "wide" | "medium" | "close" | "detail"

export type CameraMove =
  | "none"
  | "pan"
  | "tilt"
  | "orbit"
  | "push_in"
  | "pull_out"
  | "crane"

export type QuestionCategory =
  | "shot_style"
  | "pacing"
  | "camera_movement"
  | "action_beats"
  | "audio"
  | "cta"

export type ExportFormat = "json" | "markdown"

// ============================================================================
// Brief Intake (Steps 1-2)
// ============================================================================

export interface ClarifyingQuestion {
  id: string
  question: string
  category: QuestionCategory
  answer?: string
}

export interface BriefIntake {
  project_name: string
  goal: WorkflowGoal
  platform: WorkflowPlatform
  aspect_ratio: AspectRatio
  duration_seconds: number
  brand_vibe: string[]
  avoid_vibe: string[]
  must_include: string[]
  must_avoid: string[]
  script_or_vo?: string
  product_or_subject_description: string
  reference_images?: string[]
}

export interface BriefData extends BriefIntake {
  clarifying_questions: ClarifyingQuestion[]
}

// ============================================================================
// Continuity Pack (Step 3)
// ============================================================================

export interface Anchor {
  type: string
  description: string
}

export interface LightingRecipe {
  key_light: string
  fill_light: string
  rim_light?: string
  softness: string
  time_of_day: string
}

export interface PaletteAnchor {
  name: string
  hex: string
}

export interface ContinuityPack {
  anchors: Anchor[]
  lighting_recipe?: LightingRecipe
  palette_anchors: PaletteAnchor[]
  do_list: string[]
  dont_list: string[]
  style_clause?: string
}

// ============================================================================
// Shot Plan (Step 4)
// ============================================================================

export interface Shot {
  id: string
  shot_number: number
  start_sec: number
  end_sec: number
  shot_type: ShotType
  camera_move: CameraMove
  framing_notes: string
  subject_action_beats: string[]
  setting_notes: string
  audio_cue?: string
  purpose: string
}

export interface ShotPlan {
  shots: Shot[]
}

// ============================================================================
// Prompt Pack (Step 5)
// ============================================================================

export interface SoraParams {
  resolution: string
  duration: number
  aspect_ratio: string
  fps: number
  seed?: number
}

export interface ShotPrompt {
  shot_id: string
  sora_params: SoraParams
  prompt_text: string
  negative_prompt_text?: string
  dialogue_block?: string
}

export interface PromptPack {
  target_model: string
  prompts: ShotPrompt[]
}

// ============================================================================
// QA Score (Step 6)
// ============================================================================

export interface QAFix {
  issue: string
  fix: string
}

export interface QAScore {
  ambiguity_risk: number
  motion_complexity_risk: number
  continuity_completeness: number
  audio_readiness: number
  overall_score: number
  warnings: string[]
  fixes: QAFix[]
  recommended_questions: string[]
}

// ============================================================================
// Versions (Step 7)
// ============================================================================

export interface VersionSnapshot {
  version_number: number
  step: "brief" | "continuity" | "shots" | "prompts"
  data: Record<string, unknown>
  created_at: string
}

// ============================================================================
// Workflow
// ============================================================================

export interface VideoWorkflow {
  id: string
  name: string
  current_step: number
  status: WorkflowStatus
  is_template: boolean
  template_name?: string
  shot_count: number
  prompt_count: number
  overall_score?: number
  created_by?: string
  created_at: string
  updated_at: string
}

export interface VideoWorkflowDetail extends VideoWorkflow {
  source_template_id?: string
  brief?: BriefData
  continuity_pack?: ContinuityPack
  shot_plan?: ShotPlan
  prompt_pack?: PromptPack
  qa_score?: QAScore
  versions: VersionSnapshot[]
}

// ============================================================================
// API Request/Response Types
// ============================================================================

export interface CreateVideoWorkflowRequest {
  name: string
}

export interface AnswerQuestionsRequest {
  answers: Record<string, string>
}

export interface UpdateContinuityRequest {
  anchors?: Anchor[]
  lighting_recipe?: LightingRecipe
  palette_anchors?: PaletteAnchor[]
  do_list?: string[]
  dont_list?: string[]
  style_clause?: string
}

export interface UpdateShotRequest {
  shot_type?: ShotType
  camera_move?: CameraMove
  framing_notes?: string
  subject_action_beats?: string[]
  setting_notes?: string
  audio_cue?: string
  purpose?: string
  start_sec?: number
  end_sec?: number
}

export interface UpdatePromptRequest {
  prompt_text?: string
  negative_prompt_text?: string
  dialogue_block?: string
  sora_params?: SoraParams
}

export interface GeneratePromptsRequest {
  target_model?: string
}

export interface SaveVersionRequest {
  step: "brief" | "continuity" | "shots" | "prompts"
}

export interface CreateTemplateRequest {
  template_name: string
}

export interface CreateFromTemplateRequest {
  name: string
}

export interface ExportWorkflowResponse {
  format: ExportFormat
  filename: string
  content: string
}

// ============================================================================
// Step Configuration
// ============================================================================

export const WORKFLOW_STEPS = [
  { number: 1, name: "Brief Intake", key: "brief" },
  { number: 2, name: "Clarifying Questions", key: "questions" },
  { number: 3, name: "Continuity Pack", key: "continuity" },
  { number: 4, name: "Shot Plan", key: "shots" },
  { number: 5, name: "Prompt Pack", key: "prompts" },
  { number: 6, name: "QA Score", key: "qa" },
  { number: 7, name: "Export", key: "export" },
] as const

export const GOAL_OPTIONS: { value: WorkflowGoal; label: string }[] = [
  { value: "product_ad", label: "Product Ad" },
  { value: "ugc", label: "UGC / User Generated" },
  { value: "explainer", label: "Explainer Video" },
  { value: "cinematic_brand", label: "Cinematic Brand Film" },
  { value: "tutorial", label: "Tutorial / How-To" },
  { value: "other", label: "Other" },
]

export const PLATFORM_OPTIONS: { value: WorkflowPlatform; label: string }[] = [
  { value: "tiktok", label: "TikTok" },
  { value: "instagram", label: "Instagram Reels" },
  { value: "youtube", label: "YouTube" },
  { value: "web", label: "Website / Web" },
  { value: "other", label: "Other" },
]

export const ASPECT_RATIO_OPTIONS: { value: AspectRatio; label: string }[] = [
  { value: "9:16", label: "9:16 Portrait" },
  { value: "1:1", label: "1:1 Square" },
  { value: "16:9", label: "16:9 Landscape" },
]

export const DURATION_OPTIONS = [6, 12, 15, 30, 60] as const

export const SHOT_TYPE_OPTIONS: { value: ShotType; label: string }[] = [
  { value: "wide", label: "Wide" },
  { value: "medium", label: "Medium" },
  { value: "close", label: "Close-Up" },
  { value: "detail", label: "Detail / Insert" },
]

export const CAMERA_MOVE_OPTIONS: { value: CameraMove; label: string }[] = [
  { value: "none", label: "Static / Locked" },
  { value: "pan", label: "Pan" },
  { value: "tilt", label: "Tilt" },
  { value: "orbit", label: "Orbit" },
  { value: "push_in", label: "Push In / Dolly In" },
  { value: "pull_out", label: "Pull Out / Dolly Out" },
  { value: "crane", label: "Crane / Jib" },
]

File: backend/app/schemas/video_workflow.py
"""Pydantic schemas for video workflow API."""

from datetime import datetime
from enum import Enum
from typing import Optional
from pydantic import BaseModel, Field


# ============================================================================
# Enums
# ============================================================================

class WorkflowGoal(str, Enum):
    PRODUCT_AD = "product_ad"
    UGC = "ugc"
    EXPLAINER = "explainer"
    CINEMATIC_BRAND = "cinematic_brand"
    TUTORIAL = "tutorial"
    OTHER = "other"


class WorkflowPlatform(str, Enum):
    TIKTOK = "tiktok"
    INSTAGRAM = "instagram"
    YOUTUBE = "youtube"
    WEB = "web"
    OTHER = "other"


class AspectRatio(str, Enum):
    PORTRAIT = "9:16"
    SQUARE = "1:1"
    LANDSCAPE = "16:9"


class WorkflowStatus(str, Enum):
    DRAFT = "draft"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"


class ShotType(str, Enum):
    WIDE = "wide"
    MEDIUM = "medium"
    CLOSE = "close"
    DETAIL = "detail"


class CameraMove(str, Enum):
    NONE = "none"
    PAN = "pan"
    TILT = "tilt"
    ORBIT = "orbit"
    PUSH_IN = "push_in"
    PULL_OUT = "pull_out"
    CRANE = "crane"


class QuestionCategory(str, Enum):
    SHOT_STYLE = "shot_style"
    PACING = "pacing"
    CAMERA_MOVEMENT = "camera_movement"
    ACTION_BEATS = "action_beats"
    AUDIO = "audio"
    CTA = "cta"


class ExportFormat(str, Enum):
    JSON = "json"
    MARKDOWN = "markdown"


# ============================================================================
# Brief Intake (Steps 1-2)
# ============================================================================

class ClarifyingQuestion(BaseModel):
    """A clarifying question generated for the user."""
    id: str
    question: str
    category: QuestionCategory
    answer: Optional[str] = None


class BriefIntakeRequest(BaseModel):
    """Request to save/update brief intake data."""
    project_name: str = Field(..., min_length=1, max_length=255)
    goal: WorkflowGoal
    platform: WorkflowPlatform
    aspect_ratio: AspectRatio
    duration_seconds: int = Field(..., ge=3, le=180)
    brand_vibe: list[str] = Field(..., min_length=1, max_length=3)
    avoid_vibe: list[str] = Field(default_factory=list, max_length=3)
    must_include: list[str] = Field(default_factory=list)
    must_avoid: list[str] = Field(default_factory=list)
    script_or_vo: Optional[str] = None
    product_or_subject_description: str = Field(..., min_length=1, max_length=2000)
    reference_images: Optional[list[str]] = None


class AnswerQuestionsRequest(BaseModel):
    """Request to submit answers to clarifying questions."""
    answers: dict[str, str]  # question_id -> answer


class BriefData(BaseModel):
    """Full brief data including Q&A."""
    project_name: str
    goal: WorkflowGoal
    platform: WorkflowPlatform
    aspect_ratio: AspectRatio
    duration_seconds: int
    brand_vibe: list[str]
    avoid_vibe: list[str]
    must_include: list[str]
    must_avoid: list[str]
    script_or_vo: Optional[str] = None
    product_or_subject_description: str
    reference_images: Optional[list[str]] = None
    clarifying_questions: list[ClarifyingQuestion] = Field(default_factory=list)


# ============================================================================
# Continuity Pack (Step 3)
# ============================================================================

class Anchor(BaseModel):
    """A visual anchor for continuity."""
    type: str  # character, product, wardrobe, props, environment
    description: str


class LightingRecipe(BaseModel):
    """Lighting recipe for consistency."""
    key_light: str
    fill_light: str
    rim_light: Optional[str] = None
    softness: str
    time_of_day: str


class PaletteAnchor(BaseModel):
    """Color palette anchor."""
    name: str
    hex: str


class ContinuityPackData(BaseModel):
    """Full continuity pack data."""
    anchors: list[Anchor] = Field(default_factory=list)
    lighting_recipe: Optional[LightingRecipe] = None
    palette_anchors: list[PaletteAnchor] = Field(default_factory=list)
    do_list: list[str] = Field(default_factory=list)
    dont_list: list[str] = Field(default_factory=list)
    style_clause: Optional[str] = None


class UpdateContinuityRequest(BaseModel):
    """Request to update continuity pack."""
    anchors: Optional[list[Anchor]] = None
    lighting_recipe: Optional[LightingRecipe] = None
    palette_anchors: Optional[list[PaletteAnchor]] = None
    do_list: Optional[list[str]] = None
    dont_list: Optional[list[str]] = None
    style_clause: Optional[str] = None


# ============================================================================
# Shot Plan (Step 4)
# ============================================================================

class ShotData(BaseModel):
    """A single shot in the shot plan."""
    id: str
    shot_number: int
    start_sec: float
    end_sec: float
    shot_type: ShotType
    camera_move: CameraMove
    framing_notes: str
    subject_action_beats: list[str] = Field(default_factory=list, max_length=5)
    setting_notes: str
    audio_cue: Optional[str] = None
    purpose: str


class ShotPlanData(BaseModel):
    """Full shot plan data."""
    shots: list[ShotData] = Field(default_factory=list)


class UpdateShotRequest(BaseModel):
    """Request to update a single shot."""
    shot_type: Optional[ShotType] = None
    camera_move: Optional[CameraMove] = None
    framing_notes: Optional[str] = None
    subject_action_beats: Optional[list[str]] = None
    setting_notes: Optional[str] = None
    audio_cue: Optional[str] = None
    purpose: Optional[str] = None
    start_sec: Optional[float] = None
    end_sec: Optional[float] = None


# ============================================================================
# Prompt Pack (Step 5)
# ============================================================================

class SoraParams(BaseModel):
    """Sora 2 API parameters."""
    resolution: str = "1080p"
    duration: float
    aspect_ratio: str
    fps: int = 24
    seed: Optional[int] = None


class ShotPromptData(BaseModel):
    """A compiled prompt for a single shot."""
    shot_id: str
    sora_params: SoraParams
    prompt_text: str
    negative_prompt_text: Optional[str] = None
    dialogue_block: Optional[str] = None


class PromptPackData(BaseModel):
    """Full prompt pack data."""
    target_model: str = "sora_2"
    prompts: list[ShotPromptData] = Field(default_factory=list)


class UpdatePromptRequest(BaseModel):
    """Request to update a single prompt."""
    prompt_text: Optional[str] = None
    negative_prompt_text: Optional[str] = None
    dialogue_block: Optional[str] = None
    sora_params: Optional[SoraParams] = None


class GeneratePromptsRequest(BaseModel):
    """Request to generate prompts for a target model."""
    target_model: str = "sora_2"


# ============================================================================
# QA Score (Step 6)
# ============================================================================

class QAFix(BaseModel):
    """A suggested fix for a QA issue."""
    issue: str
    fix: str


class QAScoreData(BaseModel):
    """QA scoring results."""
    ambiguity_risk: int = Field(..., ge=0, le=100)
    motion_complexity_risk: int = Field(..., ge=0, le=100)
    continuity_completeness: int = Field(..., ge=0, le=100)
    audio_readiness: int = Field(..., ge=0, le=100)
    overall_score: int = Field(..., ge=0, le=100)
    warnings: list[str] = Field(default_factory=list)
    fixes: list[QAFix] = Field(default_factory=list)
    recommended_questions: list[str] = Field(default_factory=list)


# ============================================================================
# Versions (Step 7)
# ============================================================================

class VersionSnapshot(BaseModel):
    """A version snapshot of workflow data."""
    version_number: int
    step: str  # brief, continuity, shots, prompts
    data: dict
    created_at: str


class SaveVersionRequest(BaseModel):
    """Request to save a version snapshot."""
    step: str  # brief, continuity, shots, prompts


class CreateTemplateRequest(BaseModel):
    """Request to create a template from a workflow."""
    template_name: str = Field(..., min_length=1, max_length=255)


class CreateFromTemplateRequest(BaseModel):
    """Request to create a workflow from a template."""
    name: str = Field(..., min_length=1, max_length=255)


# ============================================================================
# Workflow CRUD
# ============================================================================

class CreateVideoWorkflowRequest(BaseModel):
    """Request to create a new video workflow."""
    name: str = Field(..., min_length=1, max_length=255)


class VideoWorkflowResponse(BaseModel):
    """Response containing a video workflow summary."""
    id: str
    name: str
    current_step: int
    status: WorkflowStatus
    is_template: bool
    template_name: Optional[str] = None
    shot_count: int = 0
    prompt_count: int = 0
    overall_score: Optional[int] = None
    created_by: Optional[str] = None
    created_at: str
    updated_at: str

    model_config = {"from_attributes": True}


class VideoWorkflowDetailResponse(BaseModel):
    """Response containing full video workflow details."""
    id: str
    name: str
    current_step: int
    status: WorkflowStatus
    is_template: bool
    template_name: Optional[str] = None
    source_template_id: Optional[str] = None

    # Step data
    brief: Optional[BriefData] = None
    continuity_pack: Optional[ContinuityPackData] = None
    shot_plan: Optional[ShotPlanData] = None
    prompt_pack: Optional[PromptPackData] = None
    qa_score: Optional[QAScoreData] = None
    versions: list[VersionSnapshot] = Field(default_factory=list)

    created_by: Optional[str] = None
    created_at: str
    updated_at: str

    model_config = {"from_attributes": True}


# ============================================================================
# Export
# ============================================================================

class ExportWorkflowRequest(BaseModel):
    """Request to export a workflow."""
    format: ExportFormat = ExportFormat.JSON


class ExportWorkflowResponse(BaseModel):
    """Response containing exported workflow data."""
    format: ExportFormat
    filename: str
    content: str  # JSON string or Markdown
